### ISSUE TITLE
unexpected behavior with statement_timeout

### ISSUE THREAD

**Issue by jngbng** [plv8/plv8js#439]
## Problem
When `statement_timeout` is set and plv8 function takes a long time, an exception raises at the subsequent call, not at the actual timeout  occurred.

## Reproduce scenario

```sql
begin transaction;

set local statement_timeout to 1000;

do $$
    plv8.execute('select pg_sleep(2)');
$$ language plv8;
-- it succeeds w/o an error. actually it should fail.

do $$
  return 1;
$$ language plv8;
-- any subsequent call fails with empty error!
-- ERROR:  null
-- CONTEXT:

abort;
```

If the long-running function call has made any data change, it successfully committed.
```sql
CREATE TABLE "public"."test_table" (
    "id" serial,
    "field" integer,
    PRIMARY KEY ("id")
);

begin transaction;
set local statement_timeout to 1000;
do $$
    plv8.execute('select pg_sleep(2)');
    plv8.execute('insert into test_table (field) values (1)'); 
$$ language plv8;
commit;
-- it succeeds w/o an error.

select * from test_table;
-- and can see the changes.

do $$
    return 1;
$$ language plv8;
-- again subsequent call fails with empty error!
-- ERROR:  null
-- CONTEXT:
```

## Expected behaviour

plpgsql works well with statement_timeout.

```sql
begin transaction;

set local statement_timeout to 1000;

do $$
  begin
  perform pg_sleep(2);
  end;
$$ language plpgsql;
-- ERROR:  canceling statement due to statement timeout
-- CONTEXT:  SQL statement "SELECT pg_sleep(2)"
-- PL/pgSQL function inline_code_block line 3 at PERFORM
abort;

```

## Test Environment
Test Environment1:
plv8_version: 2.3.15
postgres_version: PostgreSQL 13.1 on x86_64-apple-darwin19.6.0, compiled by Apple clang version 11.0.3 (clang-1103.0.32.62), 64-bit

Test Environment2:
plv8_version: 2.3.8
postgres_version: PostgreSQL 11.10 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-11), 64-bit

**Comment by JerrySievert** [comment]
thanks for the report!  I'll take a look at it as soon as I can.

**Comment by pkit** [comment]
@JerrySievert I think it's fixed in #441 too, but may need to add a specific tests for it (different compile flags).
I have also added timeouts to compilation phase, otherwise function compile was uninterruptable.

I think it can be backported.

**Comment by pkit** [comment]
Ah, that's a pretty peculiar case of `SPI_execute` with timeout.
It seems like it's not covered.
More than that after looking at the plpgsql code I have no idea how exactly they do it, as all the `SPI_execute*` functions are uninterruptible. And the only place where interrupts are processed is in `CHECK_FOR_INTERRUPTS()` which happens before and after execution, but not inside.
I.e. fixing the error message is easy: just add `CHECK_FOR_INTERRUPTS()` after all the SPI calls. But how to actually make it respect the timeout is unclear.

**Comment by jngbng** [comment]
This issue even crashes Postgres when plv8 function has a JSONB type parameter.

```sql
create function plv8_dummy_json(arg jsonb) returns integer as $$
  return 1;
$$ language plv8;
```

```sql
set statement_timeout to 100;
select plv8_dummy_json('[]'::jsonb) as id;
do $$ plv8.execute('select pg_sleep(1)'); $$ language plv8;
begin transaction;
select plv8_dummy_json('[]'::jsonb) as id;
-- crash here.
```

```sql
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
```

`plv8_dummy_json` does not cause crash if it does not have JSONB type parameter.

**Comment by JerrySievert** [comment]
I am unable to reproduce your (latest) issue @jngbng, can you please report which version of plv8 and which version of Postgres you are seeing this in?

**Comment by jngbng** [comment]
@JerrySievert It can be reproduced from below environments: (at least my side)

Test Environment1: (Local machine)
postgres_version: PostgreSQL 11.10 on x86_64-apple-darwin16.7.0, compiled by Apple LLVM version 8.1.0 (clang-802.0.42), 64-bit
plv8_version: 2.3.14

Test Environment2: (Local machine via Postgres.app)
postgres_version: PostgreSQL 13.1 on x86_64-apple-darwin19.6.0, compiled by Apple clang version 11.0.3 (clang-1103.0.32.62), 64-bit
plv8_version: 2.3.15

Test Environment3: (AWS RDS)
postgres_version: PostgreSQL 11.10 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-11), 64-bit
plv8_version: 2.3.8

Is it fixed in the latest plv8?

**Comment by jngbng** [comment]
Here is crash log. (I don't think it would helpful, though)

```
2021-07-30 03:01:24 UTC:(55529):[13219]:LOG:  duration: 48.908 ms  statement: create function plv8_dummy_json(arg jsonb) returns integer as $$
	  return 1;
	$$ language plv8;
2021-07-30 03:01:37 UTC:(55529):[13219]:LOG:  duration: 1006.806 ms  statement: do $$ plv8.execute('select pg_sleep(1)'); $$ language plv8;
Received signal 11 SEGV_MAPERR 000000000000

==== C stack trace ===============================

 [0x14f01755c355]
 [0x14f0438435a0]
 [0x2ed13a5907e7]
[end of stack trace]
2021-07-30 03:01:38 UTC::@:[8656]:LOG:  server process (PID 13219) was terminated by signal 11: Segmentation fault
2021-07-30 03:01:38 UTC::@:[8656]:DETAIL:  Failed process was running: select plv8_dummy_json('[]'::jsonb) as id;
2021-07-30 03:01:38 UTC::@:[8656]:LOG:  terminating any other active server processes
```

**Comment by hellower** [comment]
It is fixed in the latest plv8.

test@[local]:5432:test(12:20:25)
SQL> \dx
                                              List of installed extensions
     Name     | Version |   Schema   |                                    Description                                    
--------------+---------+------------+-----------------------------------------------------------------------------------
..........
plv8         | **3.0.0**   | pg_catalog | PL/JavaScript (v8) trusted procedural language
.........

test@[local]:5432:test(12:20:26)
SQL> create function plv8_dummy_json(arg jsonb) returns integer as $$
tarantula$#   return 1;
tarantula$# $$ language plv8;
CREATE FUNCTION
Time: 442.337 ms
test@[local]:5432:test(12:20:41)
SQL> set statement_timeout to 100;
SET
Time: 0.480 ms
test@[local]:5432:test(12:20:55)
SQL> select plv8_dummy_json('[]'::jsonb) as id;
 id 
----
  1
(1 row)

Time: 1.218 ms
test@[local]:5432:test(12:20:55)
SQL> do $$ plv8.execute('select pg_sleep(1)'); $$ language plv8;
ERROR:  Signal caught
CONTEXT:  undefined() LINE 1:  plv8.execute('select pg_sleep(1)'); 
Time: 1003.197 ms (00:01.003)
test@[local]:5432:test(12:20:56)
SQL> begin transaction;
BEGIN
Time: 0.196 ms
test@[local]:9081*:test(12:20:56)
SQL*> select plv8_dummy_json('[]'::jsonb) as id;
 id 
----
  1
(1 row)

Time: 0.744 ms
test@[local]:9081*:test(12:20:58)
SQL*> 

no crash!

**Comment by hellower** [comment]
It is fixed in plv8 3.0

```
test@[local]:5432:test(12:20:25)
SQL> \dx
                                              List of installed extensions
     Name     | Version |   Schema   |                                    Description                                    
--------------+---------+------------+-----------------------------------------------------------------------------------
..........
plv8         | 3.0.0   | pg_catalog | PL/JavaScript (v8) trusted procedural language
.........

test@[local]:5432:test(12:20:26)
SQL> create function plv8_dummy_json(arg jsonb) returns integer as $$
tarantula$#   return 1;
tarantula$# $$ language plv8;
CREATE FUNCTION
Time: 442.337 ms
test@[local]:5432:test(12:20:41)
SQL> set statement_timeout to 100;
SET
Time: 0.480 ms
test@[local]:5432:test(12:20:55)
SQL> select plv8_dummy_json('[]'::jsonb) as id;
 id 
----
  1
(1 row)

Time: 1.218 ms
test@[local]:5432:test(12:20:55)
SQL> do $$ plv8.execute('select pg_sleep(1)'); $$ language plv8;
ERROR:  Signal caught
CONTEXT:  undefined() LINE 1:  plv8.execute('select pg_sleep(1)'); 
Time: 1003.197 ms (00:01.003)
test@[local]:5432:test(12:20:56)
SQL> begin transaction;
BEGIN
Time: 0.196 ms
test@[local]:9081*:test(12:20:56)
SQL*> select plv8_dummy_json('[]'::jsonb) as id;
 id 
----
  1
(1 row)

Time: 0.744 ms
test@[local]:9081*:test(12:20:58)
SQL*> 

```
no crash!!
